File: ThreadLab4Java/ThreadLab4WithMonitor/src/Table.java
public class Table {
    private final boolean[] isEating;
    private final Object locker;

    public Table() {
        locker = new Object();
        isEating = new boolean[5];
    }

    public void takeForks(int id) {
        synchronized (locker) {
            int left = (id + 4) % 5;
            int right = (id + 1) % 5;
            while (isEating[left] || isEating[right]) {
                try {
                    locker.wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }

            isEating[id] = true;
        }
    }

    public void putForks(int id) {
        synchronized (locker) {
            isEating[id] = false;
            locker.notifyAll();
        }
    }
}

File: ThreadLab4C#/ThreadLab4/ThreadLab4WithMonitor/Table.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ThreadLab4
{
    public class Table
    {
        private readonly bool[] _isEating;
        private readonly object _locker;

        public Table()
        {
            _locker = new object();
            _isEating = new bool[5];
        }

        public void TakeForks(int id)
        {
            lock (_locker)
            {
                int left = (id + 4) % 5;
                int right = (id + 1) % 5;

                while (_isEating[left] || _isEating[right])
                {
                    Monitor.Wait(_locker);
                }

                _isEating[id] = true;
            }
        }

        public void PutForks(int id)
        {
            lock (_locker)
            {
                _isEating[id] = false;
                Monitor.PulseAll(_locker);
            }
        }
    }
}
